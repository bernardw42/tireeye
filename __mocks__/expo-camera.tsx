export const useCameraPermissions = jest.fn().mockReturnValue({
    permission: { granted: true }, // Adjust as per your test needs
    requestPermission: jest.fn().mockResolvedValue({ status: 'granted' }), // Adjust as per your test needs
  });
  