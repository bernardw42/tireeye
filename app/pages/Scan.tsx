import { CameraView, useCameraPermissions } from 'expo-camera';
import * as ImagePicker from 'expo-image-picker';
import React, { useRef, useState, useEffect } from 'react';
import { TouchableOpacity, View, Text, Button, Dimensions } from 'react-native';
import { FontAwesome5, MaterialCommunityIcons, AntDesign } from "@expo/vector-icons";
import { useNavigation } from "@react-navigation/native";

const { width, height } = Dimensions.get('window');

export default function Scan() {
  const navigation = useNavigation();
  const cameraRef = useRef<any>(null);
  const [cameraType, setCameraType] = useState<'front' | 'back'>('back');
  const [isCameraReady, setIsCameraReady] = useState(false);
  const [permission, requestPermission] = useCameraPermissions();
  const [imageUri, setImageUri] = useState<string | null>(null);

  const takePicture = async () => {
    if (cameraRef.current && isCameraReady) {
      let photo = await cameraRef.current.takePictureAsync();
      setImageUri(photo.uri);
      navigation.navigate('Info', { imageUri: photo.uri });
    }
  };

  const switchCamera = () => {
    setCameraType(
      cameraType === 'back'
        ? 'front'
        : 'back'
    );
  };

  const openImagePicker = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 3],
      quality: 1,
    });
  
    if (!result.canceled) {
      const selectedImage = result.assets[0].uri;
      setImageUri(selectedImage);
      navigation.navigate('Info', { imageUri: selectedImage });
    }
  };
  
  

  useEffect(() => {
    if (permission && !permission.granted) {
      (async () => {
        const { status } = await requestPermission();
        if (status === 'granted') {
          // Permission granted
        } else {
          // Permission denied
          console.log('Camera permission denied');
        }
      })();
    }
  }, [permission, requestPermission]);

  return (
    <View className="flex-1">
      {permission && !permission.granted ? (
        <View className="flex-1 justify-center items-center">
          <Text>We need your permission to access the camera</Text>
          <Button onPress={requestPermission} title="Grant Permission" />
        </View>
      ) : (
        <CameraView
          className="flex-1"
          facing={cameraType}
          onCameraReady={() => setIsCameraReady(true)}
          ref={cameraRef}
        >
          <View className="flex-1 relative">
            {/* Top left back arrow */}
            <View className="absolute top-0 left-0 right-0 z-10">
              <View className="absolute top-0 left-0 right-0 bg-black opacity-50 h-[300px]"></View>
              <TouchableOpacity className="pt-[50px] ml-[16px] pb-[20px] z-20" onPress={() => navigation.goBack()}>
                <AntDesign name="arrowleft" size={35} color="white" />
              </TouchableOpacity>
            </View>

            {/* Transparent black overlay with circular cutout */}
            <View className="absolute inset-0 flex items-center justify-center">
              <View
                style={{
                  width: 250,
                  height: 250,
                  borderRadius: 125,
                  borderWidth: 2,
                  borderColor: 'white',
                  backgroundColor: 'transparent',
                  position: 'absolute',
                  top: height / 2 - 125, // Center vertically
                  left: width / 2 - 125, // Center horizontally
                }}
              ></View>
            </View>

            {/* Bottom black rectangle with buttons */}
            <View className="absolute bottom-0 w-full">
              <View className="absolute bottom-0 left-0 right-0 bg-black opacity-50 h-[150px]"></View>
              <View className="flex-row justify-around w-full items-center py-10">
                <TouchableOpacity onPress={switchCamera} className="bg-white rounded-full p-5 z-30">
                  <MaterialCommunityIcons name="camera-switch" size={24} color="black" />
                </TouchableOpacity>
                <TouchableOpacity onPress={takePicture} className="bg-white rounded-full p-5 z-30">
                  <FontAwesome5 name="camera" size={24} color="black" />
                </TouchableOpacity>
                <TouchableOpacity onPress={openImagePicker} className="bg-white rounded-full p-5 z-30">
                  <MaterialCommunityIcons name="image" size={24} color="black" />
                </TouchableOpacity>
              </View>
            </View>
          </View>
        </CameraView>
      )}
    </View>
  );
};
