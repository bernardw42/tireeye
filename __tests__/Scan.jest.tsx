import * as React from 'react';
import renderer from 'react-test-renderer';
import Scan from '../app/pages/Scan';

// Mock useCameraPermissions and useNavigation
import { useCameraPermissions } from 'expo-camera';
import { useNavigation } from '@react-navigation/native';

// Mock useNavigation
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn().mockReturnValue({
    navigate: jest.fn(),
    goBack: jest.fn(),
  }),
}));

// Mock expo-camera
jest.mock('expo-camera', () => ({
  useCameraPermissions: jest.fn(),
}));

jest.mock("expo-camera")

describe('<Scan/>', () => {
  beforeEach(() => {
    const mockUseCameraPermissions = jest.fn().mockReturnValue([
      { granted: true },
      jest.fn().mockResolvedValue({ status: 'granted' }),
    ]);
    (useCameraPermissions as jest.Mock).mockImplementation(mockUseCameraPermissions);
  });

  it('has 4 children', () => {
    const tree = renderer.create(<Scan />).toJSON();
    expect(tree && tree.children.length).toBe(4); // Adjust as per your component
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Scan />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
