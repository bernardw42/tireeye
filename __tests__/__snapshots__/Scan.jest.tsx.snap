// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scan/> renders correctly 1`] = `
<View
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "justifyContent": "center",
        },
        {
          "alignItems": "center",
        },
      ]
    }
  >
    <Text>
      We need your permission to access the camera
    </Text>
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Grant Permission
        </Text>
      </View>
    </View>
  </View>
</View>
`;
